src/compiler/compiler.rs

```diff
@@ -881,33 +882,54 @@ where
     };
     let filename = filename.to_string_lossy().to_lowercase();
 
-    let rustc_vv = if filename == "rustc" || filename == "clippy-driver" {
-        // Sanity check that it's really rustc.
-        let executable = executable.to_path_buf();
-        let mut child = creator.clone().new_command_sync(executable);
-        child.env_clear().envs(ref_env(env)).args(&["-vV"]);
+    // Detect the secnario where cargo is used with sccache as a RUSTC_WRAPPER and another tool as a
+    // RUSTC_WORKSPACE_WRAPPER. In that case "rustc" will be the first argument rather than the command.
+    // As a guardrail against false positives, also check for the CARGO environment variable which
+    // cargo sets for any process it runs.
+    // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
+    let using_rustc_workspace_wrapper = env.iter().any(|(k, _)| k == OsStr::new("CARGO"))
+        && filename != "rustc"
+        && args.iter().next().map(|arg1| arg1.as_os_str()) == Some(OsStr::new("rustc"));
+
+    let rustc_vv =
+        if filename == "rustc" || filename == "clippy-driver" || using_rustc_workspace_wrapper {
+            // Sanity check that it's really rustc.
+            let executable = executable.to_path_buf();
+            let mut child = creator.clone().new_command_sync(executable);
+            // rustc workspace wrappers expect rustc as the first argument
+            if using_rustc_workspace_wrapper {
+                child.arg("rustc");
+            }
+
+            child.env_clear().envs(ref_env(env)).args(&["-vV"]);
 
-        run_input_output(child, None).await.map(|output| {
-            if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
-                if stdout.starts_with("rustc ") {
-                    return Some(Ok(stdout));
+            run_input_output(child, None).await.map(|output| {
+                if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
+                    if stdout.starts_with("rustc ") {
+                        return Some(Ok(stdout));
+                    }
                 }
-            }
-            Some(Err(ProcessError(output)))
-        })?
-    } else {
-        None
-    };
+                Some(Err(ProcessError(output)))
+            })?
+        } else {
+            None
+        };
 
     let creator1 = creator.clone();
     let executable = executable.to_owned();
-    let executable2 = executable.clone();
     let pool = pool.clone();
     let cwd = cwd.to_owned();
     match rustc_vv {
         Some(Ok(rustc_verbose_version)) => {
             debug!("Found rustc");
 
+            let executable = if using_rustc_workspace_wrapper {
+                PathBuf::from("rustc")
```

Comment:
Does it have to be `rustc`? (Related to the previous comment)

Reply:
no, I'm handling per previous comment